### PR TITLE
Add a job to check for deprecated fields

### LINKED ISSUE #282: Detecting the use of deprecated fields
To ease the upgrade from version 0 of the operator to version 1, we should have a tool that detects when someone has cluster specs that use deprecated fields and recommends newer alternatives. We may be able to do this by adding a more general action for updating the spec to replace deprecated fields with the newer alternatives, allowing the rest of the reconciliation to refer only to the latest supported fields. We could then re-use that code to build a tool that scans the clusters, runs the code to update the specs in-place, detects any changes, and prints out a new recommended spec.

We should see if there's a way to also detect fields that have changes to the defaults from the previous version and recommend that they update their specs to explicitly set the new defaults. This will allow people to opt in to new defaults earlier and avoid changes to their spec when they upgrade the operator, or to opt in to keeping the old defaults even once they're running the new version of the operator. This may get tricky with changes to highly nested fields, like the resource requirements in the pod spec, but it could still be helpful to have this feature even if it were limited to simpler fields.

### PR DESCRIPTION
**Pull request by brownleej**
Resolves #282

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -34,4 +34,7 @@ foundationdb-kubernetes-sidecar/website
 # Customizable templates
 
 config/default/manager_image_patch.yaml
-config/default/manager_image_patch.yaml-e
\ No newline at end of file
+config/default/manager_image_patch.yaml-e
+
+# General temp directory
+tmp/
```
Comment (johscheuer) [inline .gitignore:40]: NIT: could you add a newline here?

File: controllers/pod_models.go [hunk 2]
```diff
@@ -806,41 +743,262 @@ func customizeContainerFromList(containers []corev1.Container, name string, cust
...
+		spec.Processes = make(map[string]fdbtypes.ProcessSettings)
+	}
+	generalSettings := spec.Processes["general"]
+	if generalSettings.PodTemplate == nil {
+		generalSettings.PodTemplate = &corev1.PodTemplateSpec{}
+	}
+	spec.Processes["general"] = generalSettings
+}
+
+// ensureCustomParametersPresent defines custom parameters the general process
```
Comment (johscheuer) [inline controllers/pod_models.go:759]:
```suggestion
// ensureCustomParametersPresent defines custom parameters for the general process
```

File: docs/compatibility.md [hunk 3]
```diff
@@ -1,8 +1,85 @@
...
-| Operator Version  | Supported Cluster Models  | Supported FDB Versions  | Supported Kubernetes Versions |
-| ----------------- | ------------------------- | ----------------------- | ----------------------------- |
-| 0.x               | v1beta1                   | 6.1.12+                 | 1.15.0+                       |
\ No newline at end of file
+* Dropping support for old versions of FDB
+* Dropping support for old versions of Kubernetes
+* Removing deprecated fields from the cluster resource
+* Changing the default values when fields are omitted in the spec
+* Removing flags and environment variables from the operator
+* Removing behavior from the operator
```
Comment (johscheuer) [inline docs/compatibility.md:12]: We should probably also mention removing deprecated metrics?

File: docs/compatibility.md [hunk 4]
```diff
@@ -1,8 +1,85 @@
...
+running. You can find that release in the table above. This will ensure that you
+can opt in to new behavior and move to the latest supported fields in the spec
+in advance of the upgrade, through whatever process you need to update your
+clusters safely.
+
+At this point, you can run a job to check your cluster specs for deprecated
+fields or defaults. 
+
+		---
+		apiVersion: batch/v1
```
Comment (johscheuer) [inline docs/compatibility.md:42]: I think the job manifest needs some reformatting?

File: main.go [hunk 5]
```diff
@@ -110,6 +118,19 @@ func main() {
...
+		DeprecationOptions:  deprecationOptions,
+	}
+
+	if checkDeprecations {
+		go startCache(mgr)
+		err = clusterReconciler.CheckDeprecations(context.Background())
+		if err != nil {
+			setupLog.Error(err, "unable to check deprecations")
+			os.Exit(1)
+		} else {
```
Comment (johscheuer) [inline main.go:131]: The `else` can be dropped since we will exit with `1` if an error occurred.

**Review by johscheuer — COMMENTED** [review]
I just found some minor things otherwise 👍

**Review by JoshuaMcManus — APPROVED** [review]

**Pull request merged**